### PR TITLE
Rules for const safety and promotion

### PR DESCRIPTION
**Pull request by RalfJung**
Based on my blog posts and lots of prior discussion with @oli-obk and @eddyb.  I am sure I forgot some of that discussion though, so please let me know what is wrong or missing.^^

### CODE REVIEW

**Comment by RalfJung** [comment]
I significantly expanded the discussion of promotion, also covering the concerns raised in https://github.com/rust-rfcs/const-eval/issues/2.

File: const_safety.md [hunk 1]
```diff
@@ -1 +1,134 @@
...
+The miri engine, which is used to execute code at compile time, can fail in
+four possible ways:
+
+* The program performs an unsupported operation (e.g., calling an unimplemented
+  intrinsics, or doing an operation that would observe the integer address of a
+  pointer).
+* The program causes undefined behavior (e.g., dereferencing an out-of-bounds
+  pointer).
+* The program panics (e.g., a failed bounds check).
+* The program loops forever, and this is detected by the loop detector.
```
Comment (oli-obk) [inline const_safety.md]: On a best effort basis. We basically don't care too much about nontermination except for diagnostics

Comment (RalfJung) [reply const_safety.md]: AFAIK we do error hard if we detect a reoccurring state? This could happen in a promoted as well.

**Comment by RalfJung** [comment]
Note that while the `promoted.md` contains some TODOs, I think we don't have to wait merging this until the TODOs are figured out -- it's still strictly more informative than what we have now. ;)

**Pull request merged**